eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jpa/LegacyJpaEventStorageEngine.java

```diff
@@ -0,0 +1,653 @@
+            var nextSequence = aggregateSequencer.resolveBy(aggregateIdentifier).incrementAndGet();
+            return new GenericDomainEventMessage<>(
+                    aggregateType,
+                    aggregateIdentifier,
+                    nextSequence,
+                    event.getIdentifier(),
+                    event.name(),
+                    event.getPayload(),
+                    event.getMetaData(),
+                    event.getTimestamp()
+            );
+        } else {
+            // returns non-aggregate event, so the sequence is always 0
+            return new GenericDomainEventMessage<>(null,
+                                                   event.getIdentifier(),
+                                                   0L,
+                                                   event,
+                                                   event::getTimestamp);
+        }
+    }
+
+    @Override
+    public MessageStream<EventMessage<?>> source(@Nonnull SourcingCondition condition) {
+        var allCriteriaStream = condition
+                .criteria()
+                .stream()
+                .map(criteria -> this.eventsForCriteria(condition, criteria))
+                .reduce(MessageStream.empty(), MessageStream::concatWith);
+
+        var consistencyMarker = new AtomicReference<ConsistencyMarker>();
+        return allCriteriaStream.map(e -> {
+            var newMarker = consistencyMarker
+                    .accumulateAndGet(
+                            e.getResource(ConsistencyMarker.RESOURCE_KEY),
+                            (m1, m2) -> m1 == null ? m2 : m1.upperBound(m2)
+                    );
+            return e.withResource(ConsistencyMarker.RESOURCE_KEY, newMarker);
+        });
+    }
+
+    private MessageStream<EventMessage<?>> eventsForCriteria(SourcingCondition condition,
+                                                             EventCriteria criterion) {
+        var aggregateIdentifier = resolveAggregateIdentifier(criterion.tags());
+        var events = batchingOperations.readEventData(
+                aggregateIdentifier,
+                condition.start(),
+                condition.end()
+        );
+        var deserialized = upcastAndDeserializeDomainEvents(events, eventSerializer, upcasterChain);
+        return MessageStream.fromStream(
+                deserialized.asStream(),
+                e -> e,
+                LegacyJpaEventStorageEngine::fillContextWith
+        );
+    }
+
+    @Override
+    public MessageStream<EventMessage<?>> stream(@Nonnull StreamingCondition condition) {
+        var trackingToken = tokenOperations.assertGapAwareTrackingToken(condition.position());
+        var events = batchingOperations.readEventData(trackingToken);
+        var deserialized = upcastAndDeserializeTrackedEvents(events, eventSerializer, upcasterChain);
+        return MessageStream.fromStream(
+                deserialized,
+                e -> e,
```

Comment:
By using the `events` (a `Stream<TrackedEventData>`) as input, we can use the second parameter to construct an event message out of that, and the third parameter to easily extract the context values.

Reply:
I had something like that, but then I used the `upcastAndDeserializeTrackedEvents` - so looks like I need to rollback those changes :)